app/src/main/java/com/android/partagix/model/Database.kt

```diff
@@ -389,6 +392,117 @@ class Database(database: FirebaseFirestore = Firebase.firestore) {
   }
 
+  /**
+   * Retrieve all comments that a user has received, both as a lender than as an borrower
+   *
+   * @param userId the user's id
+   * @param onSuccess the function to return a List containing pairs (comment's author name,
+   *   comment)
+   */
+  fun getComments(userId: String, onSuccess: (List<Pair<String, String>>) -> Unit) {
+    val ret = mutableListOf<Pair<String, String>>()
+
+    getLoans { loans ->
+      for (loan in loans) {
+
+        if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idLender == userId // if the user is the lender,
+            &&
+            loan.reviewLender.toDouble() != 0.0 // that has been reviewed,
+            &&
+            loan.commentLender != "") { // and received a comment
+
+          getUser(loan.idBorrower, onNoUser = {}) { user ->
+            ret.add(Pair(user.name, loan.commentLender))
+          }
+        } else if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idBorrower == userId // if the user is the borrower,
+            &&
+            loan.reviewBorrower.toDouble() != 0.0 // that has been reviewed
+            &&
+            loan.commentBorrower != "") { // and received a comment
+
+          getUser(loan.idLender, onNoUser = {}) { user ->
+            ret.add(Pair(user.name, loan.commentBorrower))
+          }
+        }
+      }
+    }
+
+    onSuccess(ret)
+  }
+
+  /**
+   * Retrieve all ranks that a user has received, both as a lender than as an borrower, compute the
+   * average rank, and store it in the user's rank
+   *
+   * @param idUser the user's id
+   */
+  fun newAverageRank(idUser: String) {
+
+    var rankSum = 0.0
+    var rankCount = 0L
+
+    getLoans { loans ->
+      for (loan in loans) {
+
+        if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idLender == idUser // if the user is the lender,
+            &&
+            loan.reviewLender.toDouble() != 0.0) { // that has been reviewed
+
+          rankSum += loan.reviewLender.toDouble()
+          rankCount++
+        } else if (loan.state == LoanState.FINISHED // only finished loans
+        &&
+            loan.idBorrower == idUser // if the user is the borrower,
+            &&
+            loan.reviewBorrower.toDouble() != 0.0) { // that has been reviewed
+
+          rankSum += loan.reviewBorrower.toDouble()
+          rankCount++
+        }
+      }
+      val averageRank = rankSum / rankCount.toDouble()
+      users.document(idUser).update("rank", averageRank.toString())
+    }
+  }
+
+  /**
+   * Set a review for a loan, i.e a rank and an optional comment
+   *
+   * @param loanId the loan's id
+   * @param userId the reviewed user's id
+   * @param rank the rank to be set, must be between 0.5 and 5
+   * @param comment an optional comment to be set
+   */
+  fun setReview(
+      loanId: String,
+      userId: String,
+      rank: Double,
+      comment: String,
```

Comment:
we could have `comment: String = ""` here to have the possibility to directly call it without a comment.

Reply:
perfect